File: desktop/macos/LedgerLensDesktop/Core/StatementParser.swift
import Foundation

/// A faithful Swift port of `lib/statement-parser.js` (`parseTransactionsFromLayoutText`
/// and its helpers). The regexes, ignore rules, and control flow mirror the web/server
/// parser exactly so the desktop app produces identical rows for the same layout text.
///
/// The one difference from the server is *where the layout text comes from*: the server
/// shells out to `pdftotext -layout`; the desktop app reconstructs equivalent
/// column-preserving text on-device with PDFKit (see `PDFTextExtractor`). Nothing about a
/// user's statement ever leaves the machine.
enum StatementParser {
    // Mirrors the JS constants.
    private static let datePattern = "\\d{2}/\\d{2}/\\d{4}"
    private static let moneyToken = "(?:-|[\\d,]+\\.\\d{2}(?:CR|DR)?)"

    private static let dateRegex = try! NSRegularExpression(pattern: datePattern)

    private static let leadingNumericRegex = try! NSRegularExpression(
        pattern: "^\\d[\\d,.\\sA-Z()/-]+$"
    )

    private static let transactionLineRegex: NSRegularExpression = {
        let pattern = "^(\(datePattern))\\s+(\(datePattern))\\s+(.+?)\\s+"
            + "(\(moneyToken))\\s+(\(moneyToken))\\s+(\(moneyToken))\\s+(\(moneyToken))\\s*$"
        return try! NSRegularExpression(pattern: pattern)
    }()

    // MARK: - Public entry point

    /// Parses transaction rows from layout-preserving text, matching the server parser and
    /// capping the result at 5000 rows just like `convertStatement`.
    static func parse(layoutText text: String) -> [TransactionRow] {
        Array(parseTransactions(from: text).prefix(5000))
    }

    // MARK: - Helpers (ports of the JS functions)

    private static func normalizeWhitespace(_ value: String) -> String {
        value.replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func cleanMoney(_ value: String) -> String {
        let normalized = normalizeWhitespace(value)

        if normalized.isEmpty || normalized == "-" {
            return ""
        }

        if normalized.hasSuffix("DR") {
            return "-" + normalized.dropLast(2)
        }

        if normalized.hasSuffix("CR") {
            return String(normalized.dropLast(2))
        }

        return normalized
    }

    private static func matches(_ regex: NSRegularExpression, _ line: String) -> Bool {
        let range = NSRange(line.startIndex..<line.endIndex, in: line)
        return regex.firstMatch(in: line, range: range) != nil
    }

    private static func isIgnorableLine(_ line: String) -> Bool {
        if line.isEmpty
            || line == "\u{0c}"
            || line.hasPrefix("Page no.")
            || line.hasPrefix("Statement Summary")
            || line.hasPrefix("Please do not share your ATM")
            || line.hasPrefix("If your account is operated")
            || line.hasPrefix("This is a computer generated statement")
            || line == "Balance"
            || line == "Welcome:"
            || line.hasPrefix("Account Summary")
            || line.hasPrefix("STATEMENT OF ACCOUNT")
            || line.hasPrefix("Brought Forward")
            || line.hasPrefix("Total Debits")
            || line.hasPrefix("Total Credits")
            || line.hasPrefix("Closing Balance")
            || line.hasPrefix("anyone via email")
            || line.hasPrefix("Bank never asks for such information") {
            return true
        }

        // `(!DATE_PATTERN.test(line) && /^\d[\d,.\sA-Z()/-]+$/.test(line))`
        return !matches(dateRegex, line) && matches(leadingNumericRegex, line)
    }

    // MARK: - Core parser (port of parseTransactionsFromLayoutText)

    private struct Pending {
        let date: String
        let valueDate: String
        var descriptionParts: [String]
        let debit: String
        let credit: String
        let balance: String
    }

    private static func parseTransactions(from text: String) -> [TransactionRow] {
        let lines = text
            .replacingOccurrences(of: "\u{0c}", with: "\n")
            .components(separatedBy: "\n")
            .map { $0.replacingOccurrences(of: "\u{0000}", with: "") }
            .map { line -> String in
                // Trim trailing whitespace only (mirrors `.replace(/\s+$/g, "")`).
                var chars = Array(line)
                while let last = chars.last, last == " " || last == "\t" || last == "\r" {
                    chars.removeLast()
                }
                return String(chars)
            }

        var rows: [TransactionRow] = []
        var pending: Pending?
        var carryoverDescriptionParts: [String] = []
        var awaitingNextRowAfterPageBreak = false

        func flushPending() {
            guard let current = pending else { return }
            rows.append(
                TransactionRow(
                    id: "row-\(rows.count + 1)",
                    sourceFile: nil,
                    date: current.date,
                    valueDate: current.valueDate,
                    description: normalizeWhitespace(current.descriptionParts.joined(separator: " ")),
                    debit: current.debit,
                    credit: current.credit,
                    balance: current.balance
                )
            )
            pending = nil
        }

        for index in lines.indices {
            let rawLine = lines[index]
            let line = rawLine.trimmingCharacters(in: .whitespacesAndNewlines)

            if line.hasPrefix("Page no.") {
                flushPending()
                awaitingNextRowAfterPageBreak = true
                carryoverDescriptionParts = []
                continue
            }

            if isIgnorableLine(line) {
                continue
            }

            if let groups = capture(transactionLineRegex, line) {
                flushPending()

                pending = Pending(
                    date: groups[1],
                    valueDate: groups[2],
                    descriptionParts: carryoverDescriptionParts + [groups[3]],
                    debit: cleanMoney(groups[5]),
                    credit: cleanMoney(groups[6]),
                    balance: cleanMoney(groups[7])
                )
                carryoverDescriptionParts = []
                awaitingNextRowAfterPageBreak = false
                continue
            }

            if awaitingNextRowAfterPageBreak && pending == nil {
                carryoverDescriptionParts.append(line)
                continue
            }

            if pending == nil {
                continue
            }

            if matches(dateRegex, line) {
                flushPending()
                continue
            }

            var nextMeaningfulLine = ""
            var cursor = index + 1
            while cursor < lines.count {
                let candidate = lines[cursor].trimmingCharacters(in: .whitespacesAndNewlines)
                if candidate.isEmpty || isIgnorableLine(candidate) {
                    cursor += 1
                    continue
                }
                nextMeaningfulLine = candidate
                break
            }

            if !nextMeaningfulLine.isEmpty && matches(transactionLineRegex, nextMeaningfulLine) {
                carryoverDescriptionParts = [line]
                flushPending()
                awaitingNextRowAfterPageBreak = false
                continue
            }

            pending?.descriptionParts.append(line)
        }

        flushPending()
        return rows
    }

    /// Returns capture groups (index 0 = whole match) as strings, or nil if no match.
    private static func capture(_ regex: NSRegularExpression, _ line: String) -> [String]? {
        let range = NSRange(line.startIndex..<line.endIndex, in: line)
        guard let match = regex.firstMatch(in: line, range: range) else { return nil }

        var groups: [String] = []
        for i in 0..<match.numberOfRanges {
            let groupRange = match.range(at: i)
            if groupRange.location == NSNotFound, let swiftRange = Range(groupRange, in: line) {
                groups.append(String(line[swiftRange]))
            } else if let swiftRange = Range(groupRange, in: line) {
                groups.append(String(line[swiftRange]))
            } else {
                groups.append("")
            }
        }
        return groups
    }
}

File: desktop/windows/LedgerLens/Core/StatementParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using LedgerLens.Models;

namespace LedgerLens.Core;

/// <summary>
/// C# port of <c>lib/statement-parser.js</c> (<c>parseTransactionsFromLayoutText</c> and its
/// helpers). Regexes, ignore rules, and control flow mirror the web/server parser exactly —
/// the same logic verified byte-for-byte in the macOS app — so all platforms emit identical
/// rows for the same layout text.
/// </summary>
public static class StatementParser
{
    private const string DatePattern = @"\d{2}/\d{2}/\d{4}";
    private const string MoneyToken = @"(?:-|[\d,]+\.\d{2}(?:CR|DR)?)";

    private static readonly Regex DateRegex = new(DatePattern, RegexOptions.Compiled);
    private static readonly Regex LeadingNumericRegex = new(@"^\d[\d,.\sA-Z()/-]+$", RegexOptions.Compiled);

    private static readonly Regex TransactionLineRegex = new(
        $@"^({DatePattern})\s+({DatePattern})\s+(.+?)\s+({MoneyToken})\s+({MoneyToken})\s+({MoneyToken})\s+({MoneyToken})\s*$",
        RegexOptions.Compiled);

    private static readonly Regex WhitespaceRegex = new(@"\s+", RegexOptions.Compiled);

    /// <summary>Parses transaction rows from layout-preserving text, capped at 5000 rows.</summary>
    public static List<TransactionRow> Parse(string text)
    {
        return ParseTransactions(text).Take(5000).ToList();
    }

    private static string NormalizeWhitespace(string value) =>
        WhitespaceRegex.Replace(value, " ").Trim();

    private static string CleanMoney(string value)
    {
        var normalized = NormalizeWhitespace(value);
        if (string.IsNullOrEmpty(normalized) || normalized == "-") return "";
        if (normalized.EndsWith("DR", StringComparison.Ordinal)) return "-" + normalized[..^2];
        if (normalized.EndsWith("CR", StringComparison.Ordinal)) return normalized[..^2];
        return normalized;
    }

    private static bool IsIgnorableLine(string line)
    {
        if (string.IsNullOrEmpty(line)
            || line == "\f"
            || line.StartsWith("Page no.", StringComparison.Ordinal)
            || line.StartsWith("Statement Summary", StringComparison.Ordinal)
            || line.StartsWith("Please do not share your ATM", StringComparison.Ordinal)
            || line.StartsWith("If your account is operated", StringComparison.Ordinal)
            || line.StartsWith("This is a computer generated statement", StringComparison.Ordinal)
            || line == "Balance"
            || line == "Welcome:"
            || line.StartsWith("Account Summary", StringComparison.Ordinal)
            || line.StartsWith("STATEMENT OF ACCOUNT", StringComparison.Ordinal)
            || line.StartsWith("Brought Forward", StringComparison.Ordinal)
            || line.StartsWith("Total Debits", StringComparison.Ordinal)
            || line.StartsWith("Total Credits", StringComparison.Ordinal)
            || line.StartsWith("Closing Balance", StringComparison.Ordinal)
            || line.StartsWith("anyone via email", StringComparison.Ordinal)
            || line.StartsWith("Bank never asks for such information", StringComparison.Ordinal))
        {
            return true;
        }

        return !DateRegex.IsMatch(line) && LeadingNumericRegex.IsMatch(line);
    }

    private sealed class Pending
    {
        public string Date = "";
        public string ValueDate = "";
        public List<string> DescriptionParts = new();
        public string Debit = "";
        public string Credit = "";
        public string Balance = "";
    }

    private static List<TransactionRow> ParseTransactions(string text)
    {
        var lines = text
            .Replace('\f', '\n')
            .Split('\n')
            .Select(line => line.Replace("\0", ""))   // strip null chars (mirrors the JS null-strip)
            .Select(line => line.TrimEnd())
            .ToList();

        var rows = new List<TransactionRow>();
        Pending? pending = null;
        var carryoverDescriptionParts = new List<string>();
        var awaitingNextRowAfterPageBreak = false;

        void FlushPending()
        {
            if (pending == null) return;
            rows.Add(new TransactionRow
            {
                Id = $"row-{rows.Count + 1}",
                Date = pending.Date,
                ValueDate = pending.ValueDate,
                Description = NormalizeWhitespace(string.Join(" ", pending.DescriptionParts)),
                Debit = pending.Debit,
                Credit = pending.Credit,
                Balance = pending.Balance
            });
            pending = null;
        }

        for (var index = 0; index < lines.Count; index++)
        {
            var line = lines[index].Trim();

            if (line.StartsWith("Page no.", StringComparison.Ordinal))
            {
                FlushPending();
                awaitingNextRowAfterPageBreak = true;
                carryoverDescriptionParts = new List<string>();
                continue;
            }

            if (IsIgnorableLine(line)) continue;

            var match = TransactionLineRegex.Match(line);
            if (match.Success)
            {
                FlushPending();
                pending = new Pending
                {
                    Date = match.Groups[1].Value,
                    ValueDate = match.Groups[2].Value,
                    DescriptionParts = new List<string>(carryoverDescriptionParts) { match.Groups[3].Value },
                    Debit = CleanMoney(match.Groups[5].Value),
                    Credit = CleanMoney(match.Groups[6].Value),
                    Balance = CleanMoney(match.Groups[7].Value)
                };
                carryoverDescriptionParts = new List<string>();
                awaitingNextRowAfterPageBreak = false;
                continue;
            }

            if (awaitingNextRowAfterPageBreak && pending == null)
            {
                carryoverDescriptionParts.Add(line);
                continue;
            }

            if (pending == null) continue;

            if (DateRegex.IsMatch(line))
            {
                FlushPending();
                continue;
            }

            var nextMeaningfulLine = "";
            for (var cursor = index + 1; cursor < lines.Count; cursor++)
            {
                var candidate = lines[cursor].Trim();
                if (string.IsNullOrEmpty(candidate) || IsIgnorableLine(candidate)) continue;
                nextMeaningfulLine = candidate;
                break;
            }

            if (!string.IsNullOrEmpty(nextMeaningfulLine) && TransactionLineRegex.IsMatch(nextMeaningfulLine))
            {
                carryoverDescriptionParts = new List<string> { line };
                FlushPending();
                awaitingNextRowAfterPageBreak = false;
                continue;
            }

            pending.DescriptionParts.Add(line);
        }

        FlushPending();
        return rows;
    }
}
